File: full_frontend/js/camera.js
// Camera projection & input handling

function project(pos, camera_pos, camera_rot, focal_length = 600, zoom = 1.0) {
    const N = pos.length;
    const projected = new Array(N).fill(0).map(() => [0, 0]);
    const mask = new Array(N).fill(false);

    const yaw = camera_rot[0];
    const pitch = camera_rot[1];
    const cy = Math.cos(yaw), sy = Math.sin(yaw);
    const cp = Math.cos(pitch), sp = Math.sin(pitch);
    const eps = 1e-2;
    const f = focal_length * zoom;
    const centerX = canvas.width / 2;
    const centerY = canvas.height / 2;

    for (let i = 0; i < N; i++) {
        let px = pos[i][0] - camera_pos[0];
        let py = pos[i][1] - camera_pos[1];
        let pz = pos[i][2] - camera_pos[2];

        const x1 = cy * px + sy * py;
        const y1 = -sy * px + cy * py;
        const z1 = pz;

        const y2 = cp * y1 - sp * z1;
        const z2 = sp * y1 + cp * z1;
        const x2 = x1;

        const screen_x = centerX + f * x2 / (z2 + 5 + eps);
        const screen_y = centerY + f * y2 / (z2 + 5 + eps);

        projected[i] = [screen_x, screen_y];
        mask[i] = z2 > -4.9;
    }

    return { projected, mask };
}

// Global camera state (used by main)
let camera_rot = [0.0, 0.0];
let mouse_down = false;
let last_mouse_pos = null;
let zoom = 0.5;

function setupMouseControls(canvas) {
    canvas.addEventListener('mousedown', e => {
        if (e.button === 0) {
            mouse_down = true;
            const rect = canvas.getBoundingClientRect();
            last_mouse_pos = [e.clientX - rect.left, e.clientY - rect.top];
        }
    });

    canvas.addEventListener('mouseup', e => {
        if (e.button === 0) mouse_down = false;
    });

    canvas.addEventListener('mousemove', e => {
        if (mouse_down) {
            const rect = canvas.getBoundingClientRect();
            const x = e.clientX - rect.left;
            const y = e.clientY - rect.top;
            const dx = x - last_mouse_pos[0];
            const dy = y - last_mouse_pos[1];
            last_mouse_pos = [x, y];
            camera_rot[0] += dx * 0.005;
            camera_rot[1] += dy * 0.005;
            camera_rot[1] = Math.max(-Math.PI/2 + 0.01, Math.min(Math.PI/2 - 0.01, camera_rot[1]));
        }
    });

    canvas.addEventListener('wheel', e => {
        e.preventDefault();
        zoom *= (1.0 - Math.sign(e.deltaY) * 0.08);
        zoom = Math.max(0.05, Math.min(8.0, zoom));
    });
}

File: camera.py
import numpy as np
import pygame

def project(pos, camera_pos, camera_rot, screen_size,
            focal_length=500, zoom=1.0):
    """
    Project 3D positions into 2D screen coordinates.
    Returns:
        projected: (N,2)
        mask: particles in front of camera
        z_cam: depth values (camera space)
    """
    # Translate into camera frame
    p = pos - camera_pos
    yaw, pitch = camera_rot

    # ---------- Yaw (Z-axis) ----------
    cy, sy = np.cos(yaw), np.sin(yaw)
    x1 =  cy * p[:,0] + sy * p[:,1]
    y1 = -sy * p[:,0] + cy * p[:,1]
    z1 =  p[:,2]

    # ---------- Pitch (X-axis) ----------
    cp, sp = np.cos(pitch), np.sin(pitch)
    y2 = cp * y1 - sp * z1
    z2 = sp * y1 + cp * z1
    x2 = x1

    # ---------- Perspective ----------
    eps = 1e-2
    f = focal_length * zoom

    screen_x = screen_size / 2 + f * x2 / (z2 + 5 + eps)
    screen_y = screen_size / 2 + f * y2 / (z2 + 5 + eps)

    # Keep particles in front of camera
    mask = z2 > -4.9

    return np.column_stack((screen_x, screen_y)), mask, z2


def handle_mouse(camera_rot, mouse_down, last_mouse_pos,
                 zoom, event, sensitivity=0.005):
    """
    Mouse drag → rotate
    Mouse wheel → zoom
    """
    if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
        mouse_down = True
        last_mouse_pos = event.pos

    elif event.type == pygame.MOUSEBUTTONUP and event.button == 1:
        mouse_down = False

    elif event.type == pygame.MOUSEMOTION and mouse_down:
        dx = event.pos[0] - last_mouse_pos[0]
        dy = event.pos[1] - last_mouse_pos[1]
        last_mouse_pos = event.pos

        camera_rot[0] += dx * sensitivity
        camera_rot[1] += dy * sensitivity
        camera_rot[1] = np.clip(
            camera_rot[1],
            -np.pi/2 + 0.01,
             np.pi/2 - 0.01
        )

    elif event.type == pygame.MOUSEWHEEL:
        zoom *= (1.0 + event.y * 0.1)
        zoom = np.clip(zoom, 0.05, 6.0)

    return camera_rot, mouse_down, last_mouse_pos, zoom
